[trainer,rollout,doc] feat: reduce minimum gpus to 96 for deepseek-v3

techkang opened the pull request:
### What does this PR do?
reduce minimum gpus to 96 for deepseek-v3 and 32 for Qwen3-235B

change details:
1. use cpu adam to save GPU memory
2. change vllm sleep level to 2 to save CPU memory
3. fix conflict between megatron HybridDeviceOptimizer and verl load_megatron_optimizer.
4. provide new training scripts and document.

training logs:
DeepSeek-V3 with 12 Nodes:
<img width="3420" height="1308" alt="image" src="https://github.com/user-attachments/assets/23bec729-bf39-41c8-a4c2-c51f389d052c" />


Qwen3-235B with 4 Nodes:
<img width="3426" height="1380" alt="image" src="https://github.com/user-attachments/assets/4eeacab4-833f-4409-b294-10bd51d0fde9" />

sleep1 vs sleep2 speed on Qwen2.5-7B:
level1 mean:8.26s level2 mean: 8.33s
<img width="698" height="638" alt="image" src="https://github.com/user-attachments/assets/e3dcbb4b-f841-4d7c-b60f-40da8ffe6c42" />

chenhaiq commented on examples/grpo_trainer/run_qwen3-235b_megatron.sh:
```diff
@@ -51,13 +51,28 @@ temperature=1.0
+exp_name="235B-${NNODES}-pp${train_pp}-tp${train_tp}-ep${EP}-actor-length${actor_ppo_max_token_len}"
+CKPTS_DIR=$RAY_DATA_HOME/ckpt/${project_name}/${exp_name}
+
+# TODO: support dynamic_bsz for megatron
```
I guest you mean: "TODO: optimize cudagraph_capture_sizes for qwen3"

chenhaiq commented on examples/grpo_trainer/run_qwen3-235b_megatron.sh:
```diff
@@ -70,70 +85,94 @@ python3 -m verl.trainer.main_ppo \
     +reward_model.reward_kwargs.overlong_buffer_cfg.log=False \
     +reward_model.reward_kwargs.max_resp_len=${max_response_length} \
-    trainer.logger='["console","wandb"]' \
+    trainer.logger=['console','vemlp_wandb'] \
```
There are lots of users that can not access vemlp_wandb, you can leave it as "trainer.logger=['console']"

chenhaiq commented on verl/utils/megatron_utils.py:
```diff
@@ -491,12 +491,15 @@ def _iter_opts(opt):
-                v["exp_avg"] = v["exp_avg"].to(get_device_id(), non_blocking=True)
-            if "exp_avg_sq" in v:
-                v["exp_avg_sq"] = v["exp_avg_sq"].to(get_device_id(), non_blocking=True)
+        if hasattr(_opt.optimizer, "_move_new_state_to_right_device"):
```
please add a comment about why call this method

techkang commented on examples/grpo_trainer/run_deepseek671b_math_megatron.sh:
```diff
@@ -1,110 +1,179 @@
-set -x
```
I only have H20 GPUs and can't provide the scripts for 80GB GPUs. But users can in theory run the same script on 32 80GB GPUs only if they have enough CPU memory (>1.5 TB per node).

techkang replied:
I also renamed my script to `_96gb.sh` for clarity.

techkang commented on examples/grpo_trainer/run_qwen3-235b_megatron.sh:
```diff
@@ -1,48 +1,48 @@
 #!/usr/bin/env bash
```
ok

techkang commented:
@vermouth1992 I rebased the main branch. In the rebase I deleted the older qwen3 scripts because both scripts was tested on 96GB GPUs. Please tell me if you think we should keep it.

vermouth1992 commented:
> @vermouth1992 I rebased the main branch. In the rebase I deleted the older qwen3 scripts because both scripts was tested on 96GB GPUs. Please tell me if you think we should keep it.

No problem
